codalab/worker_manager/slurm_batch_worker_manager.py

```diff
@@ -0,0 +1,295 @@
+
+        batch_script = str(slurm_work_dir.joinpath(slurm_args['job-name'] + '.slurm'))
+        self.save_job_definition(batch_script, job_definition)
+        job_id_str = self.run_command([self.SBATCH, batch_script])
+
+        match = re.match(self.SBATCH_COMMAND_RETRUN_REGEX, job_id_str)
+        if match is not None:
+            job_id = match.group(1)
+        else:
+            logger.error("Cannot find job_id in {}.".format(job_id_str))
+            return
+
+        # Add the newly submitted job to submitted_jobs for tracking purpose
+        self.submitted_jobs.add(job_id)
+
+    def run_command(self, command):
+        proc = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        output, errors = proc.communicate()
+        if output:
+            logger.info("Executed command: {}".format(' '.join(command)))
+            return output.decode()
+        if errors:
+            print("Failed to execute {}: {}: {}".format(' '.join(command), errors, proc.returncode))
+            logger.error(errors)
+        return ""
+
+    def save_job_definition(self, job_file, job_definition):
+        """
+        Save the batch job definition to file.
+        :param job_file: a file storing the Slurm batch job configuration
+        :param job_definition: the job definition of a Slurm batch job
+        :return:
+        """
+        with open(job_file, 'w') as f:
+            f.write(job_definition)
+        logger.info("Saved the Slurm Batch Job Definition to {}".format(job_file))
+
+    def create_job_definition(self, slurm_args, command):
+        """
+        Create a Slurm batch job definition structured as a list of Slurm batch arguments and a srun command
+        :param slurm_args: arguments for launching a Slurm batch job
+        :param command: arguments for starting a CodaLab worker
+        :return: a string containing the Slurm batch job definition
+        """
+        sbatch_args = [
+            '{} --{}={}'.format(self.SBATCH_PREFIX, key, slurm_args[key])
+            for key in sorted(slurm_args.keys())
+        ]
+
+        # Check the existence of environment variables CODALAB_USERNAME and
+        # CODALAB_PASSWORD when password_file is not given.
+        worker_authentication = (
+            '[ ! -z "${CODALAB_USERNAME}" ] && [ ! -z "${CODALAB_PASSWORD}" ] || exit;\n\n'
+            if not self.args.password_file
+            else ''
+        )
+
+        # Using the --unbuffered option with srun command will allow output
+        # appear in the output file as soon as it is produced.
+        srun_args = [self.SRUN, '--unbuffered'] + command
+
+        # Job definition contains two sections: sbatch arguments and srun command
+        job_definition = (
+            '#!/usr/bin/env bash\n\n'
+            + '\n'.join(sbatch_args)
+            + '\n\n'
+            + worker_authentication
+            + ' '.join(srun_args)
+        )
+        print("Slurm Batch Job Definition")
```

Comment:
logger instead of print? (also applies to line below)

Reply:
Initially, I also thought about writing this in the logger. However, I found it doesn't look nice with the logging time in front of each line, as it breaks the bash scripts into lines. Besides, we already saved this file into local disk, add this to logger might be redundant?

Reply:
I thought the logger only adds a logging statement to the start of the first line? I think that regardless of whether it's print or logger, though, it's important to show this output—if I'm running the slurm worker manager, I don't want to have to be `cat`ting files to verify things are working, I'd rather just look at the workermanager output.

Reply:
It is currently generated in the worker manager output if you see the overview of this ticket as an example (just copied and pasted here): 
```
2020-05-20 17:45:39,729 Starting a worker!
Slurm Batch Job Definition
#!/bin/bash

#SBATCH --cpus-per-task=3
#SBATCH --gres=gpu:1
#SBATCH --job-name=stanford-codalab-slurm-worker-3086758
#SBATCH --mem-per-cpu=2048
#SBATCH --nodelist=jagupard10
#SBATCH --ntasks-per-node=1
#SBATCH --open-mode=append
#SBATCH --output=stanford-codalab-slurm-worker-3086758.out
#SBATCH --partition=jag-standard
#SBATCH --time=10-0

srun --unbuffered cl-worker --server http://localhost --verbose --exit-when-idle --idle-seconds 600 --work-dir /Users/stanford/slurm-batch-scratch/4e524789af034496b1717e943ceed111 --id 4e524789af034496b1717e943ceed111 --network-prefix cl_worker_4e524789af034496b1717e943ceed111_network --pass-down-termination
2020-05-20 17:45:39,730 Saved the Slurm....
``` 
I think what I was trying to say is that the format is a bit messy in my view, e.g. the first two lines got prefix ahead of them.
```
2020-06-01 09:55:44,590 Slurm Batch Job Definition
2020-06-01 09:55:44,590 #!/usr/bin/env bash

#SBATCH --cpus-per-task=1
#SBATCH --gres=gpu:1
#SBATCH --job-name=jingge19-codalab-slurm-worker-717845cc
#SBATCH --mem=2048
#SBATCH --nodelist=
#SBATCH --ntasks-per-node=1
#SBATCH --open-mode=append
#SBATCH --output=slurm-worker-scratch/717845cc05cb40f09a0e169cb1259d9b/jingge19-codalab-slurm-worker-717845cc.out
#SBATCH --partition=jag-standard
#SBATCH --time=10-0

srun --unbuffered cl-worker --server https://worksheets-dev.codalab.org --verbose --exit-when-idle --idle-seconds 600 --work-dir /sailhome/jingge19/slurm-codalab-worker-scratch/jingge19-717845cc05cb40f09a0e169cb1259d9b --id 717845cc05cb40f09a0e169cb1259d9b --network-prefix cl_worker_717845cc05cb40f09a0e169cb1259d9b_network --pass-down-termination --password-file /sailhome/jingge19/authentication
2020-06-01 09:55:44,592 Saved the Slurm....
```

But certainly, if you would prefer this way, I am happy to change it.